invokeai/app/services/shared/invocation_context.py

```diff
@@ -0,0 +1,625 @@
+
+            # If the invocation inherits metadata, use that. Else, use the metadata passed in.
+            metadata_ = (
+                context_data.invocation.metadata if isinstance(context_data.invocation, WithMetadata) else metadata
+            )
+
+            return services.images.create(
+                image=image,
+                is_intermediate=context_data.invocation.is_intermediate,
+                image_category=image_category,
+                board_id=board_id,
+                metadata=metadata_,
+                image_origin=ResourceOrigin.INTERNAL,
+                workflow=context_data.workflow,
+                session_id=context_data.session_id,
+                node_id=context_data.invocation.id,
+            )
+
+        def get_pil(image_name: str) -> Image:
+            """
+            Gets an image as a PIL Image object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_pil_image(image_name)
+
+        def get_metadata(image_name: str) -> Optional[MetadataField]:
+            """
+            Gets an image's metadata, if it has any.
+
+            :param image_name: The name of the image to get the metadata for.
+            """
+            return services.images.get_metadata(image_name)
+
+        def get_dto(image_name: str) -> ImageDTO:
+            """
+            Gets an image as an ImageDTO object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_dto(image_name)
+
+        def update(
+            image_name: str,
+            board_id: Optional[str] = None,
+            is_intermediate: Optional[bool] = False,
+        ) -> ImageDTO:
+            """
+            Updates an image, returning its updated DTO.
+
+            It is not suggested to update images saved by earlier nodes, as this can cause confusion for users.
+
+            If you use this method, you *must* return the image as an :class:`ImageOutput` for the gallery to
+            get the updated image.
+
+            :param image_name: The name of the image to update.
+            :param board_id: The board ID to add the image to, if it should be added.
+            :param is_intermediate: Whether the image is an intermediate. Intermediate images aren't added to the gallery.
+            """
+            if is_intermediate is not None:
+                services.images.update(image_name, ImageRecordChanges(is_intermediate=is_intermediate))
+            if board_id is None:
+                services.board_images.remove_image_from_board(image_name)
+            else:
+                services.board_images.add_image_to_board(image_name, board_id)
+            return services.images.get_dto(image_name)
+
+        self.save = save
+        self.get_pil = get_pil
+        self.get_metadata = get_metadata
+        self.get_dto = get_dto
+        self.update = update
+
+
+class LatentsInterface:
```

Comment:
I can't think of a reason for `LatentsInterface` to be part of the API if we were designing it from scratch today. Valid reasons to keep it are 1) it's too much work to move away from it, or 2) the migration would be harder for node developers.

To be explicit about it, why are we including it?

Reply:
1. > it's too much work to move away from it

I think yes, but I should be accurate and instead say I don't know how much work it would take to move away from it.

```py
@invocation_output("latents_output")
class LatentsOutput(BaseInvocationOutput):
    """Base class for nodes that output a single latents tensor"""

    tensor: torch.Tensor = PrivateAttr() # should this be omitted from the graph or held in memory throughout a graph's lifetime?
    # ... the rest of the class, but modified to not need a `LatentsField` or latents name on instantiation, somehow...

# in BaseInvocation.invoke_internal
output = self.invoke(context)
if isinstance(output, LatentsOutput):
    context.services.latents.save("some_id", output.tensor)
return output
```

Something like this is simple for an output class with a single `LatentsField`, but a bit more involved for custom outputs, which can be any arbitrary pydantic model.

We'd have to traverse the output inside `invoke_internal`, pulling out all `LatentsOutput` objects to save them.

I have a feeling there is additional complexity here that I'm missing.

2. > the migration would be harder for node developers

I'm sure we do could it in such a way that it isn't difficult to migrate, so I don't think this is a good reason.

3. We are now adding special handling for tensors. It feels wrong to do it only for tensors and not also for images. What about future non-serializable or large output types - say, a 3D model?

4. We already have a pattern for _loading_ tensors. It's awkward for one side of the API to be imperative and the other not. If we save them automatically, we should load them automatically, too. I think that would require digging deeper into `Graph`, specifically the `prepare` logic.

---

I'd like to explore this further but it may make more sense to do this in the context of a deeper rewrite.

Reply:
Sounds good to me.

When we do get around to it, I was imagining that all node outputs would just have to be serializable and we would handle them all in pretty much the same way. Some types might need custom serializers, but beyond that there would be no need for special handling of tensors.